### PR TITLE
Improve saving and loading of models

### PR DESCRIPTION
**Pull request by gonlairo**
Improve model saving in the sklearn API to ensure future compatibility and safety. Currently, we save the entire object using torch.save, which may cause issues with different CEBRA versions. This pull request (PR) tackles this problem by suggesting a solution that involves saving only the model weights and all the arguments used, allowing for complete reconstruction of the CEBRA object at loading time.

Fix https://github.com/AdaptiveMotorControlLab/CEBRA-dev/pull/643
Fix https://github.com/AdaptiveMotorControlLab/CEBRA/pull/33
Fix https://github.com/AdaptiveMotorControlLab/CEBRA-dev/issues/623

### CODE REVIEW

File: cebra/integrations/sklearn/cebra.py [hunk 1]
```diff
@@ -1191,12 +1281,25 @@ def _more_tags(self):
...
+            'device_': cebra_dict['device_'],
+            'n_features_': cebra_dict['n_features_'],
+            'n_features_in_': cebra_dict['n_features_in_'],
+            'num_sessions_': cebra_dict['num_sessions_'],
+            'offset_': cebra_dict['offset_'],
+            'solver_name_': cebra_dict['solver_name_'],
+        }
+        return state
+
+    def save(self, filename: str, backend: str = "sklearn"):
```
Comment (stes) [inline cebra/integrations/sklearn/cebra.py:1297]:
```suggestion
    def save(self, filename: str, backend: Literal["torch", "sklearn"] = "sklearn"):
```

File: cebra/integrations/sklearn/cebra.py [hunk 2]
```diff
@@ -1216,15 +1335,41 @@ def save(self, filename: str, backend: Literal["torch"] = "torch"):
...
+        else:
+            raise ValueError("CEBRA object is not fitted. "
+                             "Saving a non-fitted model is not supported.")
         return checkpoint
 
     @classmethod
     def load(cls,
              filename: str,
-             backend: Literal["torch"] = "torch",
+             backend: Literal = "auto",
```
Comment (stes) [inline cebra/integrations/sklearn/cebra.py:1372]:
```suggestion
             backend: Literal["auto", "sklearn", "torch"] = "auto",
```

File: cebra/integrations/sklearn/cebra.py [hunk 3]
```diff
@@ -12,10 +12,13 @@
 """Define the CEBRA model."""
 
 import copy
+import itertools
+import warnings
 from typing import Callable, Iterable, List, Literal, Optional, Tuple, Union
```
Comment (stes) [inline cebra/integrations/sklearn/cebra.py:17]:
```suggestion
from typing import Callable, Iterable, List, Literal, Optional, Tuple, Union, Dict
```

File: cebra/integrations/sklearn/cebra.py [hunk 4]
```diff
@@ -260,6 +263,92 @@ def _require_arg(key):
...
+        raise RuntimeError("Model loaded from file is not compatible with "
+                           "the current CEBRA version.")
+    if not sklearn_utils.check_fitted(checkpoint):
+        raise ValueError(
+            "CEBRA model is not fitted. Loading it is not supported.")
+
+    return checkpoint
+
+
+def _load_cebra_with_sklearn_backend(cebra_info: dict) -> "CEBRA":
```
Comment (stes) [inline cebra/integrations/sklearn/cebra.py:277]:
```suggestion
def _load_cebra_with_sklearn_backend(cebra_info: Dict) -> "CEBRA":
```

**Review by stes — CHANGES_REQUESTED** [review]
Some minor additions to the typing

File: cebra/integrations/sklearn/cebra.py [hunk 5]
```diff
@@ -260,6 +263,92 @@ def _require_arg(key):
...
+    """Loads a CEBRA model with a Sklearn backend.
+
+    Args:
+        cebra_info: A dictionary containing information about the CEBRA model.
```
Comment (MMathisLab) [inline cebra/integrations/sklearn/cebra.py:281]: better docstring needed; what information?

File: cebra/integrations/sklearn/cebra.py [hunk 6]
```diff
@@ -260,6 +263,92 @@ def _require_arg(key):
...
+
+    Returns:
+       The loaded CEBRA object.
+
+    Raises:
+        ValueError: If the loaded CEBRA model is not fitted, indicating that loading it is not supported.
```
Comment (MMathisLab) [inline cebra/integrations/sklearn/cebra.py:287]: grammar/Eng : is not fitted --> was not already fit

File: cebra/integrations/sklearn/cebra.py [hunk 7]
```diff
@@ -260,6 +263,92 @@ def _require_arg(key):
...
+    """
+    required_keys = ['args', 'state', 'state_dict']
+    missing_keys = [key for key in required_keys if key not in cebra_info]
+    if missing_keys:
+        raise ValueError(
+            f"Missing keys in data dictionary: {', '.join(missing_keys)}. "
+            f"You can try loading the CEBRA model with a different backend.")
```
Comment (MMathisLab) [inline cebra/integrations/sklearn/cebra.py:294]: what does this mean? what backend? What should the user actually do?

File: cebra/integrations/sklearn/cebra.py [hunk 8]
```diff
@@ -260,6 +263,92 @@ def _require_arg(key):
...
+    cebra_ = cebra.CEBRA(**args)
+
+    for key, value in state.items():
+        setattr(cebra_, key, value)
+
+    state_and_args = {**args, **state}
+
+    if not sklearn_utils.check_fitted(cebra_):
+        raise ValueError(
+            "CEBRA model is not fitted. Loading it is not supported.")
```
Comment (MMathisLab) [inline cebra/integrations/sklearn/cebra.py:307]: grammar

File: cebra/integrations/sklearn/cebra.py [hunk 9]
```diff
@@ -1205,6 +1308,22 @@ def save(self, filename: str, backend: Literal["torch"] = "torch"):
             Experimental functionality. Do not expect the save/load functionalities to be
```
Comment (MMathisLab) [inline cebra/integrations/sklearn/cebra.py:1308]: why not? this should be stable ...

File: cebra/integrations/sklearn/cebra.py [hunk 10]
```diff
@@ -1216,15 +1335,41 @@ def save(self, filename: str, backend: Literal["torch"] = "torch"):
             >>> cebra_model.save('/tmp/foo.pt')
```
Comment (MMathisLab) [inline cebra/integrations/sklearn/cebra.py:1335]: really no other inputs?

File: cebra/integrations/sklearn/cebra.py [hunk 11]
```diff
@@ -1240,6 +1385,8 @@ def load(cls,
             Experimental functionality. Do not expect the save/load functionalities to be
             backward compatible yet between CEBRA versions!
 
+            For information about the file format we refer to :py:meth:`CEBRA.save`.
```
Comment (MMathisLab) [inline cebra/integrations/sklearn/cebra.py:1388]: please refer to -->

File: cebra/integrations/sklearn/cebra.py [hunk 12]
```diff
@@ -1249,13 +1396,32 @@ def load(cls,
...
+
+        if backend == "auto":
+            backend = "sklearn" if isinstance(checkpoint, dict) else "torch"
+
+        if isinstance(checkpoint, dict) and backend == "torch":
+            raise RuntimeError(
+                f"Cannot use 'torch' backend with a dictionary-based checkpoint. "
+                f"Please try a different backend.")
+        if not isinstance(checkpoint, dict) and backend == "sklearn":
+            raise RuntimeError(f"Cannot use 'sklearn' backend. "
```
Comment (MMathisLab) [inline cebra/integrations/sklearn/cebra.py:1416]: why can't it be used? more details needed

File: tests/test_sklearn.py [hunk 13]
```diff
@@ -805,20 +808,20 @@ def _assert_same_state_dict(first, second):
             assert first[key] == second[key]
 
 
-def _assert_equal(original_model, loaded_model):
-    assert original_model.get_params() == loaded_model.get_params()
+def check_fitted(model):
```
Comment (MMathisLab) [inline tests/test_sklearn.py:811]: check_if_fit ? -- grammar!

File: tests/test_sklearn.py [hunk 14]
```diff
@@ -805,20 +808,20 @@ def _assert_same_state_dict(first, second):
...
+    """Check if a model is fitted.
```
Comment (MMathisLab) [inline tests/test_sklearn.py:812]: Check if a model has already by fit (i.e., `cebra.fit()` )

File: tests/test_sklearn.py [hunk 15]
```diff
@@ -805,20 +808,20 @@ def _assert_same_state_dict(first, second):
...
 
-    def check_fitted(model):
-        """Check if a model is fitted.
+    Args:
+        model: The model to assess.
```
Comment (MMathisLab) [inline tests/test_sklearn.py:815]: The model to check

File: tests/test_sklearn.py [hunk 16]
```diff
@@ -805,20 +808,20 @@ def _assert_same_state_dict(first, second):
...
 
-        Args:
-            model: The model to assess.
+    Returns:
+        True if fitted.
```
Comment (MMathisLab) [inline tests/test_sklearn.py:818]: grammar!

**Review by MMathisLab — CHANGES_REQUESTED** [review]

File: cebra/integrations/sklearn/cebra.py (hunk 9 above)
Comment (gonlairo) [reply cebra/integrations/sklearn/cebra.py:1311]: @stes

File: cebra/integrations/sklearn/cebra.py (hunk 10 above)
Comment (gonlairo) [reply cebra/integrations/sklearn/cebra.py:1338]: this was how the previous version behaved. we can discuss what output we should give to user

Comment (stes) [reply cebra/integrations/sklearn/cebra.py:1340]: yes, no other inputs.

**Review by stes — APPROVED** [review]
up to the remaining comment (whether to fix this as stable vs. still experimental), good to go.

**Review by MMathisLab — APPROVED** [review]

**Pull request merged**